servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/context/GraphQLContextBuilder.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2023 Expedia, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.server.execution.context
+
+import com.expediagroup.graphql.generator.extensions.toGraphQLContext
+import com.expediagroup.graphql.server.types.GraphQLServerRequest
+import graphql.GraphQLContext
+
+/**
+ * Generates a Pair that will be used as entry of a GraphQL context.
+ */
+fun interface GraphQLContextEntryProducer<Request, out K : Any, out V> {
+    suspend fun invoke(
+        request: Request,
+        graphQLRequest: GraphQLServerRequest,
+        accumulator: Map<Any, Any?>
+    ): Pair<K, V>?
+}
+
+/**
+ * Builder that takes list of [GraphQLContextEntryProducer] used to generate entries for a GraphQL context.
+ */
+interface GraphQLContextBuilder<Request> : GraphQLContextProvider<Request> {
+
+    val producers: List<GraphQLContextEntryProducer<Request, Any, *>>
+
+    override suspend fun generateContext(
+        request: Request,
+        graphQLRequest: GraphQLServerRequest,
+    ): GraphQLContext =
+        producers.fold(mutableMapOf<Any, Any?>()) { accumulator, producer ->
```

Comment:
note: while producers define their `invoke` function as `suspend`, logic below invokes it sequentially - was the intent to run them in parallel?

Reply:
given that there is an accumulator the idea is to execute them sequentially.